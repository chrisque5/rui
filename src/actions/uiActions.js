import { actionTypes } from '../utils/constants';

export function changeCounterPartySelectionAction(selectedRow) {
  return { type: actionTypes.UPDATE_COUNTERPARTY_SELECTION, selectedRow };
}

export function changeCounterPartySelection(selectedRow) {
  return (dispatch) => dispatch(changeCounterPartySelectionAction(selectedRow));
}

export function changeSelectedPreferenceBrokerAction(payload) {
  return { type: actionTypes.UPDATE_PREFERENCE_BROKER_SELECTION, payload };
}

export function changeSelectedPreferenceBroker(brokerDeskKey) {
  return (dispatch, getState) => {
    const state = getState();
    const { user: { preferences: { preferredBrokers } = {} } = {} } = state;
    const [deskId, brokerId] = brokerDeskKey.split(',');

    const payload = {
      preferredBrokers, brokerId, deskId,
    };

    dispatch(changeSelectedPreferenceBrokerAction(payload));
  };
}

export function changeLastTermSelectionAction(selectedTermId) {
  return { type: actionTypes.UPDATE_LAST_TERM_SELECTION, selectedTermId };
}

export function changeLastTermSelection(selectedTermId) {
  return (dispatch) => dispatch(changeLastTermSelectionAction(selectedTermId));
}

export function changeTermValuesAction(terms) {
  return { type: actionTypes.CHANGE_TERMS, terms };
}

export function changeTermValues(terms) {
  return (dispatch) => dispatch(changeTermValuesAction(terms));
}

export function changeStrategyTypeAction(strategy) {
  return { type: actionTypes.CHANGE_STRATEGY, strategy };
}

export function changeStrategyType(strategy) {
  return (dispatch) => dispatch(changeStrategyTypeAction(strategy));
}

export function changeDealTypeAction(dealType) {
  return { type: actionTypes.CHANGE_DEALTYPE, dealType };
}

export function changeDealType(dealType) {
  return (dispatch) => dispatch(changeDealTypeAction(dealType));
}

export function changePageAction(currentPage) {
  return { type: actionTypes.CHANGE_PAGE, currentPage };
}

export function changePage(currentPage) {
  return (dispatch) => dispatch(changePageAction(currentPage));
}

export function resetUiStateAction() {
  return { type: actionTypes.RESET_UI_STATE };
}

export function resetUiState() {
  return (dispatch) => dispatch(resetUiStateAction());
}

export function changeClientHoverDataSuccess(selectedClientData, clients) {
  return {
    type: actionTypes.CHANGE_CLIENT_HOVER_DATA_SUCCESS,
    selectedClientData,
    clients,
  };
}

export function changeClientHoverData(selectedClientData) {
  return (dispatch, getState) => {
    const { clients } = getState();
    dispatch(changeClientHoverDataSuccess(selectedClientData, clients));
  };
}

export function toggleThirdCPAction(value) {
  return { type: actionTypes.TOGGLE_THIRD_CP, value };
}

export function toggleThirdCP(value) {
  return (dispatch) => dispatch(toggleThirdCPAction(value));
}

export function toggleSettingsModalAction() {
  return { type: actionTypes.TOGGLE_SETTINGS_MODAL };
}

export function toggleSettingsModal() {
  return (dispatch) => dispatch(toggleSettingsModalAction());
}

export function toggleRenameFavoritesPopupAction(selectedFavItem) {
  return { type: actionTypes.TOGGLE_RENAME_FAVORITES_POPUP, selectedFavItem };
}

export function toggleRenameFavoritesPopup(selectedFavItem) {
  return (dispatch) => dispatch(toggleRenameFavoritesPopupAction(selectedFavItem));
}

export function toggleChangeFavoritesColourPopupAction(selectedFavItem) {
  return { type: actionTypes.TOGGLE_CHANGE_FAVORITES_COLOUR, selectedFavItem };
}

export function toggleChangeFavoritesColourPopup(selectedFavItem) {
  return (dispatch) => dispatch(toggleChangeFavoritesColourPopupAction(selectedFavItem));
}

export function enableInterestAction(value) {
  return { type: actionTypes.ENABLE_INTEREST, value };
}

export function enableInterest(value) {
  return (dispatch) => dispatch(enableInterestAction(value));
}

export function toggleTradeDateSubmitPopupAction(isVisible) {
  return { type: actionTypes.TOGGLE_TRADE_DATE_CONFIRM_POPUP, payload: { isVisible } };
}

export function toggleTradeDateSubmitPopup(isVisible) {
  return (dispatch) => dispatch(toggleTradeDateSubmitPopupAction(isVisible));
}

export function toggleIsTradeDateEnabledAction(isTradeDateEnabled) {
  return { type: actionTypes.TOGGLE_IS_TRADE_DATE_ENABLED, payload: { isTradeDateEnabled } };
}

export function toggleIsTradeDateEnabled(isTradeDateEnabled) {
  return (dispatch) => dispatch(toggleIsTradeDateEnabledAction(isTradeDateEnabled));
}

export function updateBlotterGridDetails(blotterGridUpdates) {
  return {
    type: actionTypes.UPDATE_BLOTTER_GRID_DETAILS,
    blotterGridUpdates,
  };
}

export function toggleIsClsOverrideAction(isClsOverride) {
  return { type: actionTypes.TOGGLE_IS_CLS_OVERRIDE, payload: { isClsOverride } };
}

export function toggleIsClsOverride(isClsOverride) {
  return (dispatch) => dispatch(toggleIsClsOverrideAction(isClsOverride));
}

export function resetAdminTabChangesAction(reset) {
  return { type: actionTypes.RESET_ADMIN_TAB_CHANGES, reset };
}

export function resetAdminTabChanges(reset) {
  return (dispatch) => dispatch(resetAdminTabChangesAction(reset));
}

export function enableBrowserForceRefreshAction(isBrowserForceRefreshEnabled) {
  return { type: actionTypes.ENABLE_BROWSER_FORCE_REFRESH, isBrowserForceRefreshEnabled };
}

export function enableBrowserForceRefresh(isBrowserForceRefreshEnabled) {
  return (dispatch) => dispatch(enableBrowserForceRefreshAction(isBrowserForceRefreshEnabled));
}

export function resetBlotterSearchAction(reset) {
  return { type: actionTypes.RESET_BLOTTER_SEARCH, reset };
}

export function resetBlotterSearch(reset) {
  return (dispatch) => dispatch(resetBlotterSearchAction(reset));
}

export function changeDealEditStatusAction(isDealEditInProgress) {
  return { type: actionTypes.CHANGE_DEAL_EDIT_STATUS, isDealEditInProgress };
}

export function changeDealEditStatus(isDealEditInProgress) {
  return (dispatch) => dispatch(changeDealEditStatusAction(isDealEditInProgress));
}

export function changeReAllocationAction(reAllocationSide) {
  return { type: actionTypes.CHANGE_RE_ALLOCATION_SIDE, reAllocationSide };
}

export function changeReAllocation(reAllocationSide) {
  return (dispatch) => dispatch(changeReAllocationAction(reAllocationSide));
}

export function updateDealObjectAction(updatedDealObject) {
  return { type: actionTypes.UPDATE_DEAL_OBJECT, updatedDealObject };
}

export function updateDealObject(updatedDealObject) {
  return (dispatch) => dispatch(updateDealObjectAction(updatedDealObject));
}
