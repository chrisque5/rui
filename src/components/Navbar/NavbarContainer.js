import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Modal } from 'antd';
import Navbar from './Navbar';
import * as uiActions from '../../actions/uiActions';
import * as sseActions from '../../actions/sseActions';
import * as notification from '../../utils/notifications';
import { navMenuItems, APPLICATION_TITLE } from '../../utils/constants';
import { saveCurrentPageAndRedirect } from '../../utils/windowHelper';
import { cancel } from '../../api/restApi';

class NavbarContainer extends Component {
  refreshPageAfterDateChange = () => {
    const now = new Date();
    const midNight = new Date(now.getFullYear(), now.getMonth(), now.getDate(), 23, 59, 59, 999);
    const msTillMidNight = midNight - now;
    if (msTillMidNight > 0) {
      this.refreshTimeOut = setTimeout(() => {
        this.props.uiActions.enableBrowserForceRefresh(true);
        // make sure date changes before refresh
        setTimeout(() => {
          window.location.reload(true);
        }, 100);
      },
      msTillMidNight);
    }
  };

  componentDidMount = () => {
    this.refreshPageAfterDateChange();
  };

  logout = () => {
    cancel();

    fetch(`${process.env.REACT_APP_API_URL}/services/user/expireSession`, {
      credentials: 'include',
    }).then(() => {
      notification.showSuccessNotification('Logged Out', `User ${this.props.fullName} logged out`);
    }).then(() => {
      setTimeout(() => {
        localStorage.removeItem('loggedInUser');
        // Capturing the current page to use if user login in another window
        saveCurrentPageAndRedirect(this.props.fullName);
      }, 500);
    });
  };

  confirmLogout = () => {
    Modal.confirm({
      title: 'Confirm Logout',
      content: 'Are you sure you want to logout?',
      okText: 'Logout',
      okButtonProps: { 'data-testid': 'btnLogoutConfirm' },
      cancelText: 'Cancel',
      onOk: this.logout,
    });
  }

  showSettings = () => {
    this.props.uiActions.toggleSettingsModal();
  };

  setDocumentTitle = () => {
    const navMenuItem = navMenuItems[this.props.selectedKey];
    if (navMenuItem && this.props.selectedKey !== 'EDIT') {
      document.title = `${navMenuItem.documentTitle} - ${APPLICATION_TITLE}`;
    }
  };

  componentWillUnmount = () => {
    if (this.refreshTimeOut) {
      clearTimeout(this.refreshTimeOut);
    }
  };

  onNavMenuItemChange = ({ key }) => {
    const sessionId = this.props.sessionId;
    if (navMenuItems[key] && key !== 'BLOTTER' && sessionId) {
      this.props.sseActions.terminateSSESession(sessionId);
    }
  }

  render() {
    this.setDocumentTitle();
    return (
      <Navbar
        selectedKey={this.props.selectedKey}
        fullName={this.props.fullName}
        logout={this.confirmLogout}
        settings={this.showSettings}
        permissions={this.props.permissions}
        shouldRenderLinks={this.props.shouldRenderLinks}
        onNavMenuItemChange={this.onNavMenuItemChange}
      />
    );
  }
}

NavbarContainer.defaultProps = {
  selectedKey: '',
  fullName: '',
  permissions: {},
  shouldRenderLinks: true,
  sessionId: '',
};

NavbarContainer.propTypes = {
  selectedKey: PropTypes.string,
  fullName: PropTypes.string,
  uiActions: PropTypes.shape().isRequired,
  permissions: PropTypes.objectOf(PropTypes.bool),
  shouldRenderLinks: PropTypes.bool,
  sessionId: PropTypes.string,
  sseActions: PropTypes.shape().isRequired,
};

const mapStateToProps = ({ user = {}, sse = {} }) => ({
  fullName: user.fullName,
  permissions: user.permissions,
  sessionId: sse.sessionId,
});

const mapDispatchToProps = (dispatch) => ({
  uiActions: bindActionCreators(uiActions, dispatch),
  sseActions: bindActionCreators(sseActions, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(NavbarContainer);
