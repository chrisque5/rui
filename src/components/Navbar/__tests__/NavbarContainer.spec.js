import {
  renderWithProvider, waitFor, fireEvent, getByText, getByTestId,
} from 'test-utils/testUtils';
import { HashRouter, Route } from 'react-router-dom';
import NavbarContainer from '../NavbarContainer';
import { user } from './UserData.json';
import uiActions from '../../../actions/uiActions';

global.fetch = jest.fn().mockImplementation(() => Promise.resolve({}));
jest.mock('../../../actions/uiActions', () => ({
  toggleSettingsModal: jest.fn(() => jest.fn(() => Promise.resolve(true))),
}));

const fetchSpy = jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({}));
const toggleSettingsSpy = jest.spyOn(uiActions, 'toggleSettingsModal');

const initialState = { user };

describe('<NavbarContainer />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('It renders', () => {
    const { container } = renderWithProvider(<HashRouter><Route><NavbarContainer selectedKey="FWD" /></Route></HashRouter>, { initialState });
    expect(container).toMatchSnapshot();
    expect(global.window.document.title).toBe('FWD - DMSWeb');
  });

  test('It renders correct document title', () => {
    renderWithProvider(<HashRouter><Route><NavbarContainer selectedKey="NDF" /></Route></HashRouter>, { initialState });
    expect(global.window.document.title).toBe('NDF - DMSWeb');
    renderWithProvider(<HashRouter><Route><NavbarContainer selectedKey="ADMIN" /></Route></HashRouter>, { initialState });
    expect(global.window.document.title).toBe('Admin - DMSWeb');
    renderWithProvider(<HashRouter><Route><NavbarContainer selectedKey="BLOTTER" /></Route></HashRouter>, { initialState });
    expect(global.window.document.title).toBe('Blotter - DMSWeb');
    renderWithProvider(<HashRouter><Route><NavbarContainer selectedKey="SPT" /></Route></HashRouter>, { initialState });
    expect(global.window.document.title).toBe('Spot - DMSWeb');
  });

  test.only('click Settings', async () => {
    const { container } = renderWithProvider(<HashRouter><Route><NavbarContainer selectedKey="FWD" /></Route></HashRouter>, { initialState });
    const userDropdown = await waitFor(() => getByTestId(container, 'ddlUserDropdown'));
    fireEvent.mouseOver(userDropdown);
    const btnSettings = await waitFor(() => getByTestId(container, 'btnSettings'));
    fireEvent.click(btnSettings);
    expect(toggleSettingsSpy).toHaveBeenCalledTimes(1);
  });

  test('Click Logout Cancel', async () => {
    const { container } = renderWithProvider(<HashRouter><Route><NavbarContainer selectedKey="FWD" /></Route></HashRouter>, { initialState });
    const userDropdown = await waitFor(() => getByTestId(container, 'ddlUserDropdown'));
    fireEvent.mouseOver(userDropdown);
    const btnLogout = await waitFor(() => getByTestId(document.body, 'btnLogout'));
    fireEvent.click(btnLogout);

    const cancelBtn = await waitFor(() => getByText(document.body, 'Cancel'));
    fireEvent.click(cancelBtn);
    expect(fetchSpy).toHaveBeenCalledTimes(0);
  });

  test('Click Logout Success', async () => {
    const { container } = renderWithProvider(<HashRouter><Route><NavbarContainer selectedKey="FWD" /></Route></HashRouter>, { initialState });
    const userDropdown = await waitFor(() => getByTestId(container, 'ddlUserDropdown'));
    fireEvent.mouseOver(userDropdown);
    const btnLogout = await waitFor(() => getByTestId(document.body, 'btnLogout'));
    fireEvent.click(btnLogout);

    const logoutBtnConfirm = await waitFor(() => getByTestId(document.body, 'btnLogoutConfirm'));
    fireEvent.click(logoutBtnConfirm);
    expect(fetchSpy).toHaveBeenCalledTimes(1);

    global.fetch.mockClear();
    delete global.fetch;
  });
});
