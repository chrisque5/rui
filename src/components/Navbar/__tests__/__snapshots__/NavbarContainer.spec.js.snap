// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavbarContainer /> It renders 1`] = `
<div>
  <div
    class="dms-trade-capture-navbar ant-menu-dark"
  >
    <ul
      class="ant-menu-overflow ant-menu ant-menu-root ant-menu-horizontal ant-menu-dark navbar-menu"
      data-menu-list="true"
      role="menu"
      tabindex="0"
    >
      <li
        class="ant-menu-overflow-item ant-menu-item ant-menu-item-no-link"
        data-menu-id="rc-menu-uuid-test-home"
        role="menuitem"
        style="opacity: 1; order: 0;"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <img
            alt="logo"
            height="30"
            src="tpicap.png"
            style="margin-right: 15px;"
            width="30"
          />
          <span>
            Deal Management System
          </span>
        </span>
      </li>
      <li
        aria-disabled="false"
        class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child"
        role="menuitem"
        style="opacity: 1; order: 1;"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <a
            href="#/deal/NDF"
          >
            <span>
              NDF
            </span>
          </a>
        </span>
      </li>
      <li
        class="ant-menu-overflow-item ant-menu-submenu ant-menu-submenu-horizontal user-name"
        role="none"
        style="opacity: 1; order: 2;"
      >
        <div
          aria-controls="rc-menu-uuid-test-submenu-popup"
          aria-expanded="false"
          aria-haspopup="true"
          class="ant-menu-submenu-title"
          role="menuitem"
          tabindex="-1"
        >
          <span
            class="ant-menu-title-content"
          >
            <span
              class="submenu-title-wrapper"
              data-testid="ddlUserDropdown"
              id="ddlUserDropdown"
            >
              BELFAST TPEUR TEST BROKER 1
              <span
                aria-label="caret-down"
                class="anticon anticon-caret-down"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="caret-down"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 1024 1024"
                  width="1em"
                >
                  <path
                    d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                  />
                </svg>
              </span>
            </span>
          </span>
          <i
            class="ant-menu-submenu-arrow"
          />
        </div>
      </li>
      <li
        class="ant-menu-overflow-item ant-menu-item ant-menu-item-only-child btn-settings"
        data-testid="btnSettings"
        id="btnSettings"
        role="menuitem"
        style="opacity: 1; order: 3;"
        tabindex="-1"
      >
        <span
          class="ant-menu-title-content"
        >
          <span
            aria-label="setting"
            class="anticon anticon-setting"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="setting"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512.5 390.6c-29.9 0-57.9 11.6-79.1 32.8-21.1 21.2-32.8 49.2-32.8 79.1 0 29.9 11.7 57.9 32.8 79.1 21.2 21.1 49.2 32.8 79.1 32.8 29.9 0 57.9-11.7 79.1-32.8 21.1-21.2 32.8-49.2 32.8-79.1 0-29.9-11.7-57.9-32.8-79.1a110.96 110.96 0 00-79.1-32.8zm412.3 235.5l-65.4-55.9c3.1-19 4.7-38.4 4.7-57.7s-1.6-38.8-4.7-57.7l65.4-55.9a32.03 32.03 0 009.3-35.2l-.9-2.6a442.5 442.5 0 00-79.6-137.7l-1.8-2.1a32.12 32.12 0 00-35.1-9.5l-81.2 28.9c-30-24.6-63.4-44-99.6-57.5l-15.7-84.9a32.05 32.05 0 00-25.8-25.7l-2.7-.5c-52-9.4-106.8-9.4-158.8 0l-2.7.5a32.05 32.05 0 00-25.8 25.7l-15.8 85.3a353.44 353.44 0 00-98.9 57.3l-81.8-29.1a32 32 0 00-35.1 9.5l-1.8 2.1a445.93 445.93 0 00-79.6 137.7l-.9 2.6c-4.5 12.5-.8 26.5 9.3 35.2l66.2 56.5c-3.1 18.8-4.6 38-4.6 57 0 19.2 1.5 38.4 4.6 57l-66 56.5a32.03 32.03 0 00-9.3 35.2l.9 2.6c18.1 50.3 44.8 96.8 79.6 137.7l1.8 2.1a32.12 32.12 0 0035.1 9.5l81.8-29.1c29.8 24.5 63 43.9 98.9 57.3l15.8 85.3a32.05 32.05 0 0025.8 25.7l2.7.5a448.27 448.27 0 00158.8 0l2.7-.5a32.05 32.05 0 0025.8-25.7l15.7-84.9c36.2-13.6 69.6-32.9 99.6-57.5l81.2 28.9a32 32 0 0035.1-9.5l1.8-2.1c34.8-41.1 61.5-87.4 79.6-137.7l.9-2.6c4.3-12.4.6-26.3-9.5-35zm-412.3 52.2c-97.1 0-175.8-78.7-175.8-175.8s78.7-175.8 175.8-175.8 175.8 78.7 175.8 175.8-78.7 175.8-175.8 175.8z"
              />
            </svg>
          </span>
        </span>
      </li>
      <li
        aria-hidden="true"
        class="ant-menu-overflow-item ant-menu-overflow-item-rest ant-menu-submenu ant-menu-submenu-horizontal"
        role="none"
        style="opacity: 0; height: 0px; overflow-y: hidden; order: 9007199254740991; pointer-events: none; position: absolute;"
      >
        <div
          aria-controls="rc-menu-uuid-test-rc-menu-more-popup"
          aria-expanded="false"
          aria-haspopup="true"
          class="ant-menu-submenu-title"
          data-menu-id="rc-menu-uuid-test-rc-menu-more"
          role="menuitem"
          tabindex="-1"
        >
          <span
            aria-label="ellipsis"
            class="anticon anticon-ellipsis"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="ellipsis"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z"
              />
            </svg>
          </span>
          <i
            class="ant-menu-submenu-arrow"
          />
        </div>
      </li>
    </ul>
    <div
      aria-hidden="true"
      style="display: none;"
    />
  </div>
</div>
`;
